torch/csrc/autograd/variable.h

```diff
@@ -1,150 +1,255 @@
 #pragma once
 
-// A wrapper around at::Tensor to represent autograd Variables. Variables
-// can be implicitly converted to an at::Tensor.
+#include <Python.h>
 
-#include <mutex>
-#include <memory>
-#include <vector>
-#include <functional>
-#include <ATen/ATen.h>
-
-#include "torch/csrc/assertions.h"
-#include "torch/csrc/jit/ir.h"
-#include "torch/csrc/jit/tracer_state.h"
+#include "torch/csrc/autograd/edge.h"
 #include "torch/csrc/autograd/function_hook.h"
-#include "torch/csrc/utils/auto_unique_ptr.h"
 #include "torch/csrc/autograd/variable_version.h"
-#include "torch/csrc/autograd/edge.h"
-#include "torch/csrc/Types.h"
+#include "torch/csrc/jit/tracer_state.h"
+#include "torch/csrc/utils/auto_unique_ptr.h"
+
+#include <ATen/Scalar.h>
+#include <ATen/ScalarType.h>
+#include <ATen/Storage.h>
+#include <ATen/Tensor.h>
+#include <ATen/TensorImpl.h>
+#include <ATen/Type.h>
+
+#include <list>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <vector>
 
 namespace torch { namespace autograd {
 
-using at::Tensor;
-struct VariableImpl;
+struct Function;
+
+//===----------------------------------------------------------------------===//
+//                                Variable
+//===----------------------------------------------------------------------===//
+
+/// A `Variable` augments a `Tensor` with the ability to interact in our
+/// autograd machinery. `Variable` inherits from `Tensor` and may be converted
+/// to and from `Tensor` implicitly.
+class Variable : public at::Tensor {
+ public:
+  /// Creates a Variable that is a *view* of another (*base*) variable.
+  /// The `gradient_edge` is an optional (gradient_function, input_number) pair.
+  static Variable
+  as_view(Variable base, at::Tensor data, Edge gradient_edge = Edge());
+
+  Variable() = default;
+  Variable(at::Tensor data, bool requires_grad);
+  Variable(at::Tensor data, Edge gradient_edge);
+
+  // "Downcasts" a `Tensor` into a `Variable`. Only call this on tensors you
+  // know are Variables.
+  /*implicit*/ Variable(at::Tensor const& rhs) : at::Tensor(rhs) {}
+  /*implicit*/ Variable(at::Tensor&& rhs) noexcept
+      : at::Tensor(std::move(rhs)) {}
+
+  // NOTE: Assignment operators to Tensor come for free from the constructors.
+
+  /// Compare this `Variable` to another `Variable` (or `Tensor`) via
+  /// pointer-equality.
+  bool operator==(const Variable& other) const noexcept {
+    return this->pImpl == other.pImpl;
+  }
+
+  bool operator!=(const Variable& other) const noexcept {
+    return !(*this == other);
+  }
```

Comment:
I don't think that's a great idea. Not sure what this overload does in ATen, but I'd expect `tensor1 == tensor2` to give me back a bool tensor of the same size, which tells me what elements are equal. Making it different for `Variables` would be very inconsistent

Reply:
Ok. It replaces code that was 'tensor1.get() == tensor1.get()'. I can make it `is_equal()`.

Reply:
`is_equal` looks to much like `bool equal(const Tensor & other)` to me, which compares values. Maybe `is_same`?

Reply:
Ok